src/eddymotion/model/dipy.py

```diff
@@ -0,0 +1,209 @@
+
+    Parameters
+    ----------
+    kernel_model: :obj:`str`
+        The string representing the desired kernel model.
+
+    Returns
+    -------
+    :obj:`GaussianProcessRegressor.kernel`
+        A GaussianProcessRegressor kernel object.
+
+    Raises
+    ------
+    TypeError: If the provided kernel_model is not supported.
+
+    """
+
+    if kernel_model == 'spherical':
+        raise NotImplementedError("Spherical kernel is not currently implemented.")
+
+    if kernel_model == 'exponential':
+        raise NotImplementedError("Exponential kernel is not currently implemented.")
+
+    if kernel_model == 'test':
+        from sklearn.gaussian_process.kernels import DotProduct, WhiteKernel
+
+        return DotProduct() + WhiteKernel()
+
+    raise TypeError(f"Unknown kernel '{kernel_model}'.")
+
+
+class GaussianProcessModel(ReconstModel):
+    """A Gaussian Process (GP) model to simulate single- and multi-shell DWI data."""
+
+    __slots__ = (
+        "kernel_model",
+        "mask",
+        "_gpr",
+    )
+
+    def __init__(
+        self,
+        gtab: GradientTable,
+        kernel_model: str = "spherical",
+        random_state: int = 0,
+        *args,
+        **kwargs,
+    ):
+        """A GP-based DWI model [Andersson15]_.
+
+        Parameters
+        ----------
+        gtab : GradientTable class instance
+
+        kernel_model : str
+            Kernel model to calculate the GP's covariance matrix.
+
+        References
+        ----------
+        .. [Andersson15] Jesper L.R. Andersson and Stamatios N. Sotiropoulos.
+           Non-parametric representation and prediction of single- and multi-shell
+           diffusion-weighted MRI data using Gaussian processes. NeuroImage, 122:166-176, 2015.
+           doi:\
+           `10.1016/j.neuroimage.2015.07.067 <https://doi.org/10.1016/j.neuroimage.2015.07.067>`__.
+
+        """
+
+        ReconstModel.__init__(self, gtab)
+        self.kernel_model = kernel_model
+        self.gtab = gtab
+        self._gpr = GaussianProcessRegressor(
+            kernel=get_kernel(self.kernel_model),
+            random_state=random_state,
+        )
+
+    def fit(self, data, gtab=None, mask=None):
+        """Fit method of the DTI model class
+
+        Parameters
+        ----------
+        data : array
+            The measured signal from one voxel.
+
+        mask : array, optional
+            A boolean array used to mark the coordinates in the data that
+            should be analyzed that has the shape data.shape[:-1]
+
+        """
+
+        if mask is not None:
+            data = data[mask[..., None]]
+        else:
+            data = np.reshape(data, (-1, data.shape[-1]))
+
+        gtab = gtab if gtab is not None else self.gtab
+        return GPFit(
+            self._gpr.fit(gtab, data),
+            gtab=gtab,
+            mask=mask,
+        )
+
+    @multi_voxel_fit
+    def multi_fit(self, data_thres, mask=None, **kwargs):
```

Comment:
What is the purpose of this method? Also, looks like the mask doesn't get used below.

Reply:
I separated the decorated fit because if I interpret the `multi_voxel_fit` decorator properly, an array with shape (n_parameters, n_voxels) is expected as the output of the decorated function, so that then parameters can be stacked/manipulated and used in prediction.

In our case, instead, we get a trained GaussianProcessRegressor object instance. We can probably extract the parameters that describe it, but I don't know how to do that at this point.

So I moved it out to punt the implementation at a later time when everything is working for a single thread single voxel fit.

WDTY?